File: python/year2021/day21.py
import time
import os
import collections
import itertools


# ------ Part 1 ------ #
class Player:
    def __init__(self, start_position):
        self.position = start_position
        self.score = 0


class DeterministicDice:
    def __init__(self):
        self.current_value = 0
        self.rolls = 0

    def roll(self):
        self.current_value += 1
        self.current_value %= 100
        self.rolls += 1
        return self.current_value


def parse_input(filename: str) -> (list[Player], int, int):
    with open(filename) as file:
        players = []
        for line in file:
            players.append(Player(int(line.split(":")[1]) - 1))
    return players, players[0].position, players[1].position


def do_round_pt1(player: Player, dice: DeterministicDice) -> bool:
    rolls = (dice.roll(), dice.roll(), dice.roll())
    player.position += sum(rolls)
    player.position %= 10
    player.score += player.position + 1  # The board is numbered from 1-10, but I'm storing 0-9.
    return player.score >= 1000


def play_game_pt1(players: list[Player], dice: DeterministicDice):
    while True:
        for i, player in enumerate(players):
            if do_round_pt1(player, dice):
                return i


# ------ Part 2 ------ #
PlayerState = collections.namedtuple("PlayerState", ["p1_pos", "p1_score", "p2_pos", "p2_score", "current_player"])
Win = collections.namedtuple("Win", ["player"])  # mostly just to give a more specific name than "int"


def find_next_state(player_state: PlayerState):
    states: list[PlayerState or Win] = []

    for rolls in itertools.product(range(1, 4), repeat=3):
        pos = (player_state.p1_pos if player_state.current_player == 0 else player_state.p2_pos)
        pos += sum(rolls)
        pos %= 10
        score = (player_state.p1_score if player_state.current_player == 0 else player_state.p2_score) + 1 + pos
        if score >= 21:
            states.append(Win(player_state.current_player))
        elif player_state.current_player == 0:
            states.append(PlayerState(pos, score, player_state.p2_pos, player_state.p2_score, 1))
        elif player_state.current_player == 1:
            states.append(PlayerState(player_state.p1_pos, player_state.p1_score, pos, score, 0))
    return states


def play_game_pt2(p1_start, p2_start):
    states = dict()
    quantities = collections.Counter()
    quantities[PlayerState(p1_start, 0, p2_start, 0, 0)] = 1
    p1_wins = 0
    p2_wins = 0
    while sum(quantities.values()) > 0:
        next_quantities = collections.Counter()
        for start_state in quantities:
            # This check saves having to work out states that can't actually happen, which evidentally seems
            # to be a sizable number since it more than halves the run time to do it this way compared to
            # checking every state that seems even remotely maybe possible.
            if start_state not in states:
                states[start_state] = find_next_state(start_state)
            for state in states[start_state]:
                if type(state) == Win:
                    if state.player == 0:
                        p1_wins += quantities[start_state]
                    elif state.player == 1:
                        p2_wins += quantities[start_state]
                else:  # type(state) == PlayerState
                    next_quantities[state] += quantities[start_state]
        quantities = next_quantities
    return max(p1_wins, p2_wins), min(p1_wins, p2_wins), int(p1_wins < p2_wins), int(p1_wins > p2_wins)


def main(input_filename: str):
    start_time = time.time()
    players, p1_start, p2_start = parse_input(input_filename)

    part1_start = time.time()
    dice = DeterministicDice()
    loser = (play_game_pt1(players, dice) + 1) % 2  # only works because there's only two players.
    print(f"Part 1: Loser's score {players[loser].score} * number of rolls {dice.rolls} = ", end="")
    print(players[loser].score * dice.rolls)

    part2_start = time.time()
    winner_wins, loser_wins, winner_id, loser_id = play_game_pt2(p1_start, p2_start)
    print(f"Part 2: The winner, player {winner_id + 1}, wins in {winner_wins} universes")
    print(f"Extra: The loser, player {loser_id + 1}, wins in {loser_wins} universes")

    end_time = time.time()
    print("Elapsed Time:")
    print(f"    Parsing: {(part1_start - start_time) * 1000:.2f} ms")
    print(f"    Part 1: {(part2_start - part1_start) * 1000:.2f} ms")
    print(f"    Part 2: {(end_time - part2_start) * 1000:.2f} ms")
    print(f"    Total: {(end_time - start_time) * 1000:.2f} ms")


if __name__ == "__main__":
    os.chdir(os.path.split(__file__)[0])
    main("../../inputs/2021/day21.txt")

File: cpp/source/2021/day21.cpp
#include <iostream>
#include <sstream>
#include <fstream>
#include <array>
#include <unordered_map>
#include <numeric>
#include <fmt/format.h>
#include <dstr/string.h>
#include "../days.h"

// ----- PART 1 -----
struct Player
{
	int32_t position;
	int32_t score;
};

class DeterministicDice
{
public:
	DeterministicDice() : current_value(0), rolls(0) {}
	int32_t current_value;
	int32_t rolls;
	int32_t roll()
	{
		++current_value;
		current_value %= 100;
		++rolls;
		return current_value;
	}
};

std::array<Player, 2> parse_input(const std::string& filename)
{
	auto file = std::ifstream(filename);
	std::stringstream contents;
	contents << file.rdbuf();
	auto lines = dstr::split(dstr::strip(contents.str()), "\n");
	std::array<Player, 2> players{};
	for (size_t i = 0; i < lines.size(); ++i) {
		std::cout << fmt::format("Current line: {}\n", lines[i]);
		players[i] = { std::stoi(dstr::split(lines[i], ":")[1]) - 1, 0 };
	}
	return players;
}

bool do_round_pt1(Player& player, DeterministicDice& dice)
{
	for (int i = 0; i < 3; ++i) {
		auto roll = dice.roll();
		player.position += roll;
	}
	player.position %= 10;
	player.score += player.position + 1;
	return player.score >= 1000;
}

size_t play_game_pt1(std::array<Player, 2>& players, DeterministicDice& dice)
{
	while (true) {
		for (size_t i = 0; i < players.size(); ++i) {
			if (do_round_pt1(players[i], dice)) return i;
		}
	}
}

// ----- PART 2 -----
struct PlayerState
{
	int32_t p1_pos;
	int32_t p1_score;
	int32_t p2_pos;
	int32_t p2_score;
	int32_t current_player;

	bool operator==(const PlayerState& other) const noexcept {
		return p1_pos == other.p1_pos
			&& p1_score == other.p1_score
			&& p2_pos == other.p2_pos
			&& p2_score == other.p2_score
			&& current_player == other.current_player;
	}
};

template<>
struct std::hash<PlayerState>
{
	std::size_t operator()(const PlayerState& s) const noexcept
	{
		return static_cast<size_t>(s.p1_pos | (s.p1_score << 4) | (s.p2_pos << 9) | (s.p2_score << 13) | (s.current_player << 18));
	}
};

std::vector<PlayerState> find_next_state(const PlayerState& state) 
{
	std::vector<PlayerState> states;
	for (int32_t i = 1; i < 4; ++i) {
		for (int32_t j = 1; j < 4; ++j) {
			for (int32_t k = 1; k < 4; ++k) {
				int32_t pos = state.current_player == 0 ? state.p1_pos : state.p2_pos;
				pos += i + j + k;
				pos %= 10;
				int32_t score = (state.current_player == 0 ? state.p1_score : state.p2_score) + 1 + pos;
				if (state.current_player == 0) {
					states.emplace_back(PlayerState{ pos, score, state.p2_pos, state.p2_score, 1 });
				}
				else { // (state.current_player == 1)
					states.emplace_back(PlayerState{ state.p1_pos, state.p1_score, pos, score, 0 });
				}
			}
		}
	}
	return states;
}

bool is_empty(const std::unordered_map<PlayerState, int64_t>& umap)
{
	for (const auto& entry : umap) {
		if (entry.second > 0) return false;
	}
	return true;
}

void play_game_pt2(int32_t p1_start, int32_t p2_start)
{
	constexpr int32_t WIN = 21;
	std::unordered_map<PlayerState, std::vector<PlayerState>> states;
	std::unordered_map<PlayerState, int64_t> quantities;
	quantities[PlayerState{ p1_start, 0, p2_start, 0, 0 }] = 1;
	int64_t p2_wins = 0;
	int64_t p1_wins = 0;
	while (!is_empty(quantities)) {
		std::unordered_map<PlayerState, int64_t> next_quantities;
		for (const auto& entry : quantities) {
			auto& start_state = entry.first;
			auto& quantity = entry.second;
			if (states.find(start_state) == states.end()) {
				states[start_state] = find_next_state(start_state);
			}
			for (const auto& state : states[start_state]) {
				if (state.p1_score >= WIN) p1_wins += quantities[start_state];
				else if (state.p2_score >= WIN) p2_wins += quantities[start_state];
				else next_quantities[state] += quantities[start_state];
			}
		}
		std::swap(quantities, next_quantities);
	}
	auto win_score = std::max(p1_wins, p2_wins);
	auto lose_score = std::min(p1_wins, p2_wins);
	int winner = p1_wins < p2_wins;
	int loser = p1_wins > p2_wins;
	std::cout << fmt::format("Part 2: The winner, player {}, wins in {} universes\n", winner, win_score);
	std::cout << fmt::format("Extra: The loser, player {}, wins in {} universes\n", loser, lose_score);
}

std::optional<std::chrono::duration<float>> run_2021_day_21(const std::string& filename)
{
	auto start_time = std::chrono::system_clock::now();
	auto players = parse_input(filename);
	auto p1_start = players[0].position;
	auto p2_start = players[1].position;

	auto part1_start = std::chrono::system_clock::now();
	auto dice = DeterministicDice();
	auto loser = (play_game_pt1(players, dice) + 1) % 2;
	std::cout << fmt::format(
		"Part 1: Loser's score {} * number of rolls {} = {}\n",
		players[loser].score,
		dice.rolls,
		players[loser].score * dice.rolls
	);

	auto part2_start = std::chrono::system_clock::now();
	// Note: To simplify returns here, Part 2's functions will handle printing results.
	play_game_pt2(p1_start, p2_start);
	auto end_time = std::chrono::system_clock::now();

	typedef std::chrono::duration<float> time;
	std::cout << "Elapsed Time:\n";
	std::cout << fmt::format("    Parsing: {:.03} ms\n", time(part1_start - start_time).count() * 1000);
	std::cout << fmt::format("    Part 1: {:.03} ms\n", time(part2_start - part1_start).count() * 1000);
	std::cout << fmt::format("    Part 2: {:.03} ms\n", time(end_time - part2_start).count() * 1000);
	time elapsed_time = (end_time - start_time);
	std::cout << fmt::format("    Total: {:.03} ms\n", elapsed_time.count() * 1000);
	return std::make_optional(elapsed_time);
}